File: friedman/little-java/compare/Expression.java
public final class Expression {
    static class BadResultException extends RuntimeException {
        private static final long serialVersionUID = -7471855055854681068L;
        BadResultException(String s) { super(s); }
    }

    interface Exp {
        Exp eval();
        boolean hasZero();
        @Override
        String toString();
    }

    static class Int implements Exp {
        final int value;
        Int(int value) { this.value = value; }

        @Override
        public Exp eval() {
            return this;
        }
        @Override
        public boolean hasZero() {
            return value == 0;
        }
        @Override
        public String toString() {
            return String.valueOf(value);
        }
    }

    static class Negate implements Exp {
        final Exp exp;
        Negate(Int exp) { this.exp = exp; }

        @Override
        public Exp eval() {
            Int intExp = (Int) exp.eval();
            try {
                return new Int(-intExp.value);
            } catch (ClassCastException cce) {
                throw new BadResultException("non-int in negation");
            }
        }
        @Override
        public boolean hasZero() {
            return exp.hasZero();
        }
        @Override
        public String toString() {
            return "-(" + exp.toString() + ")";
        }
    }

    static class Add implements Exp {
        final Exp e1;
        final Exp e2;
        Add(Exp e1, Exp e2) { this.e1 = e1; this.e2 = e2; }

        @Override
        public Exp eval() {
            try {
                Int i1 = (Int) e1.eval();
                Int i2 = (Int) e2.eval();
                return new Int(i1.value + i2.value);
            } catch (ClassCastException e) {
                throw new BadResultException("non-ints in addition");
            }
        }
        @Override
        public boolean hasZero() {
            return e1.hasZero() || e2.hasZero();
        }
        @Override
        public String toString() {
            return "(" + e1.toString() + ") + (" + e2.toString() + ")";
        }
    }

    static class Mult implements Exp {
        final Exp e1;
        final Exp e2;
        Mult(Exp e1, Exp e2) { this.e1 = e1; this.e2 = e2; }

        @Override
        public Exp eval() {
            try {
                Int i1 = (Int) e1.eval();
                Int i2 = (Int) e2.eval();
                return new Int(i1.value * i2.value);
            } catch (ClassCastException e) {
                throw new BadResultException("non-ints in addition");
            }
        }
        @Override
        public boolean hasZero() {
            return e1.hasZero() || e2.hasZero();
        }
        @Override
        public String toString() {
            return "(" + e1.toString() + ") * (" + e2.toString() + ")";
        }
    }

    public static void main(String[] args) {
        System.out.println(new Add(new Negate(new Int(5)), new Int(6)).eval());
    }
}

File: friedman/little-java/compare/expression.ml
exception BadResult of string

type exp =
    Int    of int
  | Negate of exp
  | Add    of exp * exp

let rec eval = function  e ->
  match e with
      Int _      -> e
    | Negate e1  -> (match eval e1 with
                        Int i -> Int (-i)
                      | _ -> raise (BadResult "non-int in negation"))
    | Add(e1,e2) -> (match (eval e1, eval e2) with
                        (Int i, Int j) -> Int (i+j)
                      | _ -> raise (BadResult "non-ints in addition"))

let rec toString = function e ->
  match e with
      Int i      -> string_of_int i
    | Negate e1  -> "-(" ^ (toString e1) ^ ")"
    | Add(e1,e2) -> "("  ^ (toString e1) ^ " + " ^ (toString e2) ^ ")"

let rec hasZero = function e ->
  match e with
      Int i      -> i=0
    | Negate e1  -> hasZero e1
    | Add(e1,e2) -> (hasZero e1) || (hasZero e2)

(* let rec noNegConstants = function e ->
  match e with
      Int i       -> if i < 0 then Negate(Int(-i)) else e
    | Negate e1   -> Negate(noNegConstants e1)
    | Add(e1,e2)  -> Add(noNegConstants e1, noNegConstants e2) *)

(* type ’a ext_exp =
    Int    of int
  | Negate of ’a ext_exp
  | Add    of ’a ext_exp * ’a ext_exp
  | OtherExtExp  of ’a

let rec eval_ext = function (f, e) ->
  match e with
      Int i -> i
    | Negate e1 -> 0 - (eval_ext (f,e1))
    | Add(e1,e2) -> (eval_ext (f,e1)) + (eval_ext (f,e2)) 
    | OtherExtExp e -> f e *)
;;
toString (eval (Add ((Negate (Int 5)), (Int 6))))